components/tooltip/Tooltip.js

```diff
@@ -0,0 +1,271 @@
+          events.removeEventListenerOnTransitionEnded(this.tooltipNode, this.onTransformEnd);
+        }
+
+        if (this.timeout) {
+          clearTimeout(this.timeout);
+        }
+      }
+
+      onTransformEnd = e => {
+        if (e.propertyName === 'transform') {
+          events.removeEventListenerOnTransitionEnded(this.tooltipNode, this.onTransformEnd);
+          this.setState({ visible: false });
+        }
+      };
+
+      getPosition(element) {
+        const { tooltipPosition } = this.props;
+
+        if (tooltipPosition === POSITION.HORIZONTAL) {
+          const origin = element.getBoundingClientRect();
+          const { width: ww } = getViewport();
+          const toRight = origin.left < ww / 2 - origin.width / 2;
+
+          return toRight ? POSITION.RIGHT : POSITION.LEFT;
+        } else if (tooltipPosition === POSITION.VERTICAL) {
+          const origin = element.getBoundingClientRect();
+          const { height: wh } = getViewport();
+          const toBottom = origin.top < wh / 2 - origin.height / 2;
+
+          return toBottom ? POSITION.BOTTOM : POSITION.TOP;
+        }
+
+        return tooltipPosition;
+      }
+
+      activate({ top, left, position }) {
+        if (this.timeout) {
+          clearTimeout(this.timeout);
+        }
+
+        this.setState({ visible: true, position });
+
+        this.timeout = setTimeout(() => {
+          this.setState({ active: true, top, left });
+        }, this.props.tooltipDelay);
+      }
+
+      deactivate() {
+        if (this.timeout) {
+          clearTimeout(this.timeout);
+        }
+
+        if (this.state.active) {
+          events.addEventListenerOnTransitionEnded(this.tooltipNode, this.onTransformEnd);
+          this.setState({ active: false });
+        } else if (this.state.visible) {
+          this.setState({ visible: false });
+        }
+      }
+
+      calculatePosition(element) {
+        const { top, left, height, width } = element.getBoundingClientRect();
+        const position = this.getPosition(element);
+        const xOffset = window.scrollX || window.pageXOffset;
+        const yOffset = window.scrollY || window.pageYOffset;
+
+        if (position === POSITION.BOTTOM) {
+          return {
+            top: top + height + yOffset,
+            left: left + width / 2 + xOffset,
+            position,
+          };
+        } else if (position === POSITION.TOP) {
+          return {
+            top: top + yOffset,
+            left: left + width / 2 + xOffset,
+            position,
+          };
+        } else if (position === POSITION.LEFT) {
+          return {
+            top: top + height / 2 + yOffset,
+            left: left + xOffset,
+            position,
+          };
+        } else if (position === POSITION.RIGHT) {
+          return {
+            top: top + height / 2 + yOffset,
+            left: left + width + xOffset,
+            position,
+          };
+        }
+
+        return undefined;
```

Comment:
You don't really have to explicitly return `undefined`. When a function does not return anything, it returns `undefined` by default.